### ISSUE TITLE
Have kfree()-family have a NULL pointer-setting side-effect

### ISSUE THREAD

**Issue by kees** [KSPP/linux#87]
It would be nice to eliminate this redundancy in the kernel:

```
kfree(ptr);
ptr = NULL;
```

and instead, make it implicit when calling `kfree()`, so that when storage is freed, the pointer to it is also reset to avoid misusing these kinds of stored but freed pointers.

**Comment by alexdewar** [comment]
I'm guessing that you were thinking we could implement this with a macro?

I guess another option would be to change the argument to ``kfree()`` to ``void **`` to make this more explicit, but changing the API might be a bit horrible.

Let me know if you'd like help with this :-)

**Comment by gwelymernans** [comment]
Something like this?

```
void __kfree(void *ptr);
#define kfree(ptr)  __kfree((ptr) ? &(ptr) : NULL)
```

**Comment by alexdewar** [comment]
@gwelymernans No, ``kfree()`` already checks for null inputs. The idea would be to set the value of ``ptr`` itself to null so that it's not left dangling.

**Comment by allenpais** [comment]
Well, the simplest I can think of is something like(obviously it can't be used in it current state):
void kfree_safe(void **ptr)
{
    if (ptr && *ptr) {
        free(ptr);
        ptr = NULL;
    }
}

Since it has to be acceptable upstream, I need to do a better job.

**Comment by kees** [comment]
Without doing massive tree-wide changes, I think we need compiler support. If we had something like `__builtin_is_lvalue()`, we could distinguish function returns from lvalues. For example, right now a common case are things like:

```
kfree(get_some_ptr());
```

But if we could at least gain coverage of the lvalue cases, and detect them statically at compile-time, we could do:

```
#define __kfree_and_null(x) do { __kfree(*x); *x = NULL; } while (0)
#define kfree(x) __builtin_choose_expr(__builtin_is_lvalue(x), __kfree_and_null(&(x)), __kfree(x))
```

Alternatively, we could do a tree-wide change of the former case (findable with Coccinelle) and change them into something like `kfree_no_null()` and redefine `kfree()` itself:

```
#define kfree_no_null(x) do { void *__ptr = (x); __kfree(__ptr); } while (0)
#define kfree(x) do { __kfree(x); x = NULL; } while (0)
```

I suspect anything needing treewide changes will be met with grumpiness. :)

**Comment by kees** [comment]
https://github.com/kees/llvm-project/commits/builtin_is_lvalue/
https://lore.kernel.org/linux-hardening/20250321202620.work.175-kees@kernel.org/
